File: backend/app/models/analysis.py
"""심층 분석 ★ (구현스펙 08절, S8). 판정은 권고이지 결정이 아니다 — cite(조문 위치) 없는 판정은
화면에 내보내지 않는다는 CLAUDE.md 원칙이 analysis_requirement.cite에 그대로 반영됨.
"""

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Table,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB

from app.models.base import metadata

analysis = Table(
    "analysis",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("notice_id", Integer, ForeignKey("notice.id")),  # url/upload 입력은 nullable
    Column("source_kind", String(10), nullable=False),  # notice/url/upload
    Column("input_ref", Text, nullable=False),
    Column("status", String(20), nullable=False, server_default="queued"),  # queued/running/done/failed
    Column("step", String(50)),
    Column("started_at", DateTime(timezone=True)),
    Column("finished_at", DateTime(timezone=True)),
    Column("verdict", Text),
    Column("confidence", Numeric(4, 3)),
    Column("llm_tokens", Integer, nullable=False, server_default="0"),
    Column("llm_cost", Numeric(10, 4), nullable=False, server_default="0"),  # USD, A2·A5·A6 호출 누계
    # A2가 요구사양(analysis_requirement)과 별개로 뽑는 서술형 요약 — 사업개요·사업내용·평가기준처럼
    # 항목/값/연산자로 쪼갤 수 없는 내용(2026-09-05, 공고 상세페이지 종합 요청). 판정이 아니라
    # 추출·정리이므로 원칙 1과 무관.
    Column("summary", JSONB, nullable=True),
    Column("created_by", String(255), nullable=False, server_default="report@grib.co.kr"),
    Column("ver", Integer, nullable=False, server_default="1"),  # 동일 공고 재분석 시 버전으로 쌓임
    Column("created_at", DateTime(timezone=True), nullable=False, server_default=func.now()),
)

analysis_doc = Table(
    "analysis_doc",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("analysis_id", Integer, ForeignKey("analysis.id", ondelete="CASCADE"), nullable=False),
    Column("name", String(255), nullable=False),
    Column("kind", String(10), nullable=False),  # hwp/hwpx/pdf/html
    Column("bytes", Integer, nullable=False),
    Column("sha256", String(64), nullable=False),
    Column("extract_method", String(30)),  # 폴백 사슬 중 성공한 단계 — 조용한 빈 결과 금지
    Column("extract_ok", Boolean, nullable=False, server_default="false"),
    Column("error", Text),
    # 원본 파일은 저장하지 않는다(2026-09-03 결정) — notice.url이 항상 있어 필요하면 재다운로드
    # 가능하고, 저장공간이 무한정 느는 것도 피한다. 추출된 텍스트만 여기 남긴다.
    Column("text", Text),
)

analysis_requirement = Table(
    "analysis_requirement",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("analysis_id", Integer, ForeignKey("analysis.id", ondelete="CASCADE"), nullable=False),
    Column("category", String(50), nullable=False),
    Column("req_text", Text, nullable=False),
    Column("req_value", String(100)),
    Column("req_unit", String(20)),
    Column("op", String(10), nullable=False),  # gte/lte/eq/contains/manual
    Column("cite", Text, nullable=False),  # 규격서 조문 위치 — 근거 없는 판정 금지(S8 원칙 2)
    # 이 요구사항이 특정 과제(analysis.summary.content_items[].title)에 속하면 그 title을 그대로
    # 담는다(2026-09-05, "성능 요구사항은 과제 개요의 각 과제 하위에" 요청) — 화면에서 과제별로
    # 묶어 보여주기 위한 용도일 뿐, 판정 로직(match.py)엔 영향 없음.
    Column("task_ref", String(200)),
    Column("matched_product_id", Integer, ForeignKey("product.id")),
    Column("judgement", String(10), nullable=False, server_default="unknown"),  # ok/no/unknown
    Column("note", Text),
)

analysis_flag = Table(
    "analysis_flag",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("analysis_id", Integer, ForeignKey("analysis.id", ondelete="CASCADE"), nullable=False),
    Column("title", String(200), nullable=False),
    Column("quote", Text, nullable=False),
    Column("judgement", String(20), nullable=False),
    Column("action", Text),
)

analysis_check = Table(
    "analysis_check",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("analysis_id", Integer, ForeignKey("analysis.id", ondelete="CASCADE"), nullable=False),
    Column("title", String(200), nullable=False),
    Column("detail", Text),
    Column("critical", Boolean, nullable=False, server_default="false"),
    Column("done", Boolean, nullable=False, server_default="false"),
    Column("done_by", String(255)),
    Column("done_at", DateTime(timezone=True)),
)

# 사내 sLLM(B, extract-requirements) 요구사항 추출 미리보기(2026-09-20, 의사결정_로그 178번).
# analysis_requirement(A2, Haiku)와 물리적으로 분리된 테이블이다 — match.py(A3)는 이 테이블을
# 절대 읽지 않는다. sLLM 추출은 A2만큼 검증되지 않은 무료 미리보기일 뿐, 실제 제품 매칭
# 판정의 근거가 되면 안 된다(두 계층 LLM 설계: sLLM=미리보기, Haiku=확정).
analysis_sllm_preview = Table(
    "analysis_sllm_preview",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("analysis_id", Integer, ForeignKey("analysis.id", ondelete="CASCADE"), nullable=False, unique=True),
    Column("status", String(20), nullable=False),  # sLLM job 상태 값 그대로: queued/running/done/failed
    Column("sllm_job_id", String(100)),
    Column("chunks_processed", Integer),
    Column("chunks_total", Integer),
    # sllm_verification.sanitize_sllm_requirements()를 거친 결과만 저장(근거 검증+중복 제거 후)
    Column("requirements", JSONB),
    Column("rejected_ungrounded_count", Integer),
    Column("duplicate_count", Integer),
    Column("error", Text),
    Column("started_at", DateTime(timezone=True), nullable=False, server_default=func.now()),
    Column("finished_at", DateTime(timezone=True)),
)

File: frontend/src/api/analysis.ts
import { apiClient } from "@/api/client";

// S8 파일럿(A0+A1만, 2026-09-03) — 첨부문서 텍스트 추출까지만. LLM 판정(A2 이후)은 범위 밖.
export interface AnalysisDoc {
  name: string;
  kind: string; // pdf/hwpx/hwp/other
  extract_method: string | null;
  extract_ok: boolean;
  error: string | null;
  text: string | null;
}

export interface ExtractionResult {
  analysis_id: number;
  status: string; // running/done/failed
  step?: string;
  ver?: number;
  finished_at?: string | null;
  attachments_found?: number;
  docs: AnalysisDoc[];
  error?: string | null; // status가 failed일 때 사유(2026-09-06 — 조용히 실패하지 않도록 화면에 노출)
}

export async function fetchLatestExtraction(noticeId: number): Promise<ExtractionResult | null> {
  const { data } = await apiClient.get<ExtractionResult | null>(`/notices/${noticeId}/extract`);
  return data;
}

export async function runExtraction(noticeId: number): Promise<ExtractionResult> {
  const { data } = await apiClient.post<ExtractionResult>(`/notices/${noticeId}/extract`);
  return data;
}

// S8 A2(요구사양 구조화, LLM, 2026-09-05) — 판정(A3)은 아직 없다. judgement 필드가 없는 건
// 의도된 것 — "unknown"이라도 보이면 마치 충족 여부가 정해진 것처럼 오해를 줄 수 있어 백엔드
// 응답 자체에서 뺐다.
export interface Requirement {
  category: string;
  req_text: string;
  req_value: string | null;
  req_unit: string | null;
  op: "gte" | "lte" | "eq" | "contains" | "manual";
  cite: string;
  task_ref: string | null; // 성능 요구사항이 특정 과제(content_items[].title)에 속하면 그 title
}

export interface AnalysisEvaluationItem {
  item: string;
  weight: string;
  note?: string;
}

export interface AnalysisBudgetConditions {
  government_support_ratio: string;
  institution_cash_burden_ratio: string;
  tech_fee_collection: string;
  youth_hiring_requirement: string;
  labor_cost_basis: string;
}

export interface AnalysisTaskType {
  execution_system: string; // 추진체계
  development_form: string; // 개발형태
  call_type: string; // 공모형태
}

export interface AnalysisContact {
  department: string;
  role: string;
  phone: string;
  email: string;
}

export interface AnalysisContentItem {
  title: string;
  summary: string;
  period: string;
  budget: string;
  task_type: AnalysisTaskType; // 과제마다 추진체계/개발형태/공모형태가 다를 수 있음(2026-09-05)
  lead_org: string; // 이 과제만의 주관연구개발기관 제한(예: "비영리기관") — 참여 판단에 결정적일 수 있음
}

export interface AnalysisEligibility {
  consortium: string;
  lead_org: string;
  participant_org: string;
  demand_org: string;
  company_size: string;
  special_notes: string;
}

export interface AnalysisSubmission {
  deadline: string;
  method: string;
  documents: string[];
}

export interface AnalysisSummary {
  project_period: string;
  project_budget: string;
  purpose: string;
  sub_business: string;
  task_type: AnalysisTaskType;
  contact: AnalysisContact;
  content_items: AnalysisContentItem[];
  evaluation: AnalysisEvaluationItem[];
  budget_conditions: AnalysisBudgetConditions;
  eligibility: AnalysisEligibility;
  submission: AnalysisSubmission;
  other_notes: string;
}

export interface RequirementsResult {
  analysis_id: number;
  status: string;
  step: string | null;
  summary: AnalysisSummary | null;
  // true면 summary가 "진짜 최신 분석"이 아니라 그 이전에 AI분석까지 완료된 버전에서 가져온
  // 값이다 — 첨부문서만 다시 추출(재추출)하고 아직 AI분석은 안 돌린 상태(2026-09-13).
  summary_outdated: boolean;
  requirements: Requirement[];
}

export type LlmModel = "haiku" | "sonnet" | "opus";

export async function fetchRequirements(noticeId: number): Promise<RequirementsResult | null> {
  const { data } = await apiClient.get<RequirementsResult | null>(`/notices/${noticeId}/requirements`);
  return data;
}

export async function runStructuring(noticeId: number, model: LlmModel): Promise<{ saved: number; cost_usd: number }> {
  const { data } = await apiClient.post(`/notices/${noticeId}/structure`, { model });
  return data;
}

// 사내 sLLM(A, extract-requirements) 요구사항 추출 미리보기(2026-09-20) — A2(Haiku) 실행 전
// 무료로 먼저 훑어보는 용도. analysis_requirement(A2 확정 결과)와 별개이며 판정 근거로 쓰지
// 않는다 — "확인 필요" 미리보기로만 노출한다.
export interface SllmRequirementPreview {
  status: "queued" | "running" | "done" | "failed";
  chunks_processed: number | null;
  chunks_total: number | null;
  requirements: Requirement[] | null;
  rejected_ungrounded_count: number | null;
  duplicate_count: number | null;
  error: string | null;
}

export async function startSllmRequirementPreview(noticeId: number): Promise<SllmRequirementPreview> {
  const { data } = await apiClient.post<SllmRequirementPreview>(`/notices/${noticeId}/sllm-requirements`);
  return data;
}

export async function fetchSllmRequirementPreview(noticeId: number): Promise<SllmRequirementPreview | null> {
  const { data } = await apiClient.get<SllmRequirementPreview | null>(`/notices/${noticeId}/sllm-requirements`);
  return data;
}
